### PR TITLE
Make UNET training optional

### PR DESCRIPTION
**Pull request by leppie**
This is still a working progress, but it makes unet training optional.

This allows one to train in multiple steps when you dont have enough VRAM.

For example, where training the UNET and Text encoder is not possible on a 10GB VRAM GPU, this can get you close to that by training the Text Encoder first and then the UNET. I only have a 10GB 3080 and currently I can achieve quite good results in < 20 minutes of training on a subject. 

There is still a lot of testing to be done to find out what are good steps and learning rates to use.

This feature is NOT user friendly at this point of time.

### CODE REVIEW

**Comment by leppie** [comment]
## Current settings

This working for me on a 10GB 3080 on Win10, with Chrome (50+ tabs), Visual Studio 2022, SSMS 19, VSCode, Excel all open. 

Subject images: 8
Class images: 800 (very random and poor quality, but seems to work well)

### Step 1:
```
***** Running training *****
  Num examples = 8
  Num batches each epoch = 8
  Num Epochs = 300
  Instantaneous batch size per device = 1
  Total train batch size (w. parallel, distributed & accumulation) = 8
  Gradient Accumulation steps = 8
  Total optimization steps = 300
  Actual steps: 300
   Training settings: CPU: False, Adam: True, Prec: fp16, Grad: True, TextTr: True, EM: False, LR: 2e-06, LORA: False, UNET: False
```
Peaks around 8.8GB

### Step 2:
```
***** Running training *****
  Num examples = 8
  Num batches each epoch = 8
  Num Epochs = 100
  Instantaneous batch size per device = 1
  Total train batch size (w. parallel, distributed & accumulation) = 2
  Gradient Accumulation steps = 2
  Total optimization steps = 400
  Actual steps: 400
   Training settings: CPU: False, Adam: True, Prec: fp16, Grad: True, TextTr: False, EM: False, LR: 1e-06, LORA: False, UNET: True
```
Peaks around 9.0GB

**Comment by leppie** [comment]
## Current settings (only what has changed from above):

### Step 1: 
Epochs 150, Linear LR of 0.0001

### Step 2:
Epochs 50, Linear LR of 0.0001

Training wall time: 7:15 minutes (from clicking Train, with 2x saving checkpoints, 4x images samples, and settings changeover)
Training time: 1:49 + 3:43

**Comment by Anime4000** [comment]
I can confirm this works on my 3080 10GB

if I could automatically run in batches:
1. Text Encoder = True, UNET = False
2. Text Encoder = False, UNET = True

**Comment by leppie** [comment]
Thanks @Anime4000 for an option.

**Comment by Anime4000** [comment]
Good news to RTX 3080 10GB Users

10GB GPUs can train **UNET** and **Text Encoder** separately and automatically!
Stage 1
Stage 2

If you want to use this change on your installation, just open Console/Terminal at `extensions/sd_dreambooth_extension` folder then execute this command `gh pr checkout 540` (require [GitHub CLI](https://cli.github.com/) to be installed)

If Stable Diffusion still running, restart it (close and reopen), then do:
image

- [x] Use UNET
- [x] Use 8bit Adam
- [x] Train UNET and Text Encoder Separately
- [x] Train Text Encoder
- [x] Mixed Precision: `fp16`
- [x] Memory Attention: `xformers`

⚠️ When you load previous training, make sure double check all parameter before start training!

Good Luck ♥

**Comment by leppie** [comment]
Good start but a couple of teething issues.

* If you cancel in step 1, step 2 still goes on.
* Cant resume correctly now. 

Will see if I can address these issues to be more streamlined.

File: dreambooth/train_dreambooth.py [hunk 1]
```diff
@@ -553,6 +553,7 @@ def create_vae():
 
     unet_lora_params = None
     text_encoder_lora_params = None
+    args.pad_tokens = args.train_text_encoder
```
Comment (leppie) [inline dreambooth/train_dreambooth.py:556]: This slipped in, but been meaning to get a PR out for this

**Comment by leppie** [comment]
Added a slider now to control the ratio. 

Gradio is not playing well with it.

Also adjust gradient steps when doing step 1, so it can be done faster.

**Comment by d8ahazard** [comment]
For the UI stuff, I think it'd make more sense to change it from "Use UNET" to "Train UNET", and put the "Train Text Encoder" setting between it and "Train in 2 stages" or whatever it's called.

Also, in the start_training method, you should add a check to the others to ensure that at least the unet or text encoder are enabled for training (or lora, idk if this affects that), but basically, a sanity check to ensure that *something* is going to be trained.

**Comment by leppie** [comment]
Bad bot! Reopen
